libraries/bot-builder/src/main/java/com/microsoft/bot/builder/CloudAdapterBase.java

```diff
@@ -0,0 +1,456 @@
+    ) {
+        // Set the callerId on the activity.
+        activity.setCallerId(authenticateRequestResult.getCallerId());
+
+        // Create the connector client to use for outbound requests.
+        return authenticateRequestResult.getConnectorFactory().create(
+            activity.getServiceUrl(),
+            authenticateRequestResult.getAudience())
+            .thenCompose(connectorClient -> {
+            // Create a UserTokenClient instance for the application to use. (For example, it would be used in a sign-in prompt.)
+            BotFrameworkAuthentication.createUserTokenClient(authenticateRequestResult.getClaimsIdentity())
+                .thenCompose(userTokenClient -> {
+                    // Create a turn context and run the pipeline.
+                    TurnContext context = createTurnContext(
+                        activity,
+                        authenticateRequestResult.getClaimsIdentity(),
+                        authenticateRequestResult.getAudience(),
+                        connectorClient,
+                        userTokenClient);
+
+                    // Run the pipeline
+                    runPipeline(context, callbackHandler).thenApply(task -> {
+                        // If there are any results they will have been left on the TurnContext.
+                        return CompletableFuture.completedFuture(processTurnResults(context));
+                    });
+            });
+        });
+    }
+
+    /**
+     * This is a helper to create the ClaimsIdentity structure from an appId that will be added to the TurnContext.
+     * It is intended for use in proactive and named-pipe scenarios.
+     * @param botAppId The bot's application id.
+     * @return A {@link ClaimsIdentity} with the audience and appId claims set to the appId.
+     */
+    protected ClaimsIdentity createClaimsIdentity(String botAppId) {
+        if (botAppId == null) {
+            botAppId = "";
+        }
+
+        // Hand craft Claims Identity.
+        HashMap<String, String> claims = new HashMap<String, String>();
+        // Adding claims for both Emulator and Channel.
+        claims.put(AuthenticationConstants.AUDIENCE_CLAIM, botAppId);
+        claims.put(AuthenticationConstants.APPID_CLAIM, botAppId);
+        ClaimsIdentity claimsIdentity = new ClaimsIdentity("anonymous", claims);
+
+        return claimsIdentity;
+    }
+
+    private TurnContext createTurnContext(
+        Activity activity,
+        ClaimsIdentity claimsIdentity,
+        String oauthScope,
+        ConnectorClient connectorClient,
+        RestOAuthClient userTokenClient,
+        BotCallbackHandler callback,
+        ConnectorFactory connectorFactory
+    ) {
+        TurnContext turnContext = new TurnContextImpl(this, activity);
+        turnContext.getTurnState().add(BotAdapter.BOT_IDENTITY_KEY, claimsIdentity);
+        turnContext.getTurnState().add(BotFrameworkAdapter.CONNECTOR_CLIENT_KEY, connectorClient);
```

Comment:
Strange but I suppose there's good reason for this (the key I mean)

Reply:
Yes, this is something that can confirm during the validation. The `CloudAdapterBase` is also getting the `connectorClient` using that key, but in C# it's never assigned.

I added that key but we need to confirm this